### PR TITLE
refactor: 1:1 mapping for `ActionCosts` to fees

### PR DESCRIPTION
**Pull request by jakmeier**
So far, several actions fees were conflated right when they are added to
the profile. Here is a first step to make profiles universal by tracking
in more fine-grained manner at the profile interface. Inside, it's
still conflated but it's easier to refactor in steps. (see #8033)

### CODE REVIEW

**Comment by jakmeier** [comment]
Context for next steps as outline in #8033 is quite important here. Notably, the `DataProfile` is about to be frozen and only used to read old profiles from DB. So the fixed length hack is only temporary.

The reason for this intermediate step is some circular requirements. I can't easily change the profile's inside before changing runtime config to use parameters everywhere. But I cannot use parameters for actions as long as there is no 1:1 mapping.
Further, using parameters requires to change how the protocol config is serialized on the RPC.

File: core/primitives-core/src/profile.rs [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
+/// Serialization format to store profiles in the database.
+///
+/// This is not part of the protocol but archival nodes still rely on this not
+/// changing to answer old tx-status requests with a gas profile.
 #[derive(Clone, PartialEq, Eq)]
 pub struct DataArray(Box<[u64; Self::LEN]>);
 
 impl DataArray {
-    pub const LEN: usize = Cost::ALL.len();
+    pub const LEN: usize = if cfg!(feature = "protocol_feature_ed25519_verify") { 72 } else { 70 };
```
Comment (matklad) [inline core/primitives-core/src/profile.rs:15]: Let's add a test that the `LEN` is correct, by comparing it with `iter().count()`

File: core/primitives-core/src/profile.rs [hunk 2]
```diff
@@ -91,41 +94,33 @@ impl ProfileData {
...
-        let mut host_gas = 0u64;
-        for cost in Cost::ALL {
-            match cost {
-                Cost::ExtCost { ext_cost_kind: e } => host_gas += self.get_ext_cost(*e),
-                _ => {}
-            }
-        }
-        host_gas
+    fn host_gas(&self) -> u64 {
+        ExtCosts::iter().map(|a| self.get_ext_cost(a)).sum()
```
Comment (matklad) [inline core/primitives-core/src/profile.rs:110]: While wee are here, let's maybe saturating_add ? Seem better to not think about overflows in profiles at all

File: core/primitives/src/views.rs [hunk 3]
```diff
@@ -1244,28 +1245,53 @@ impl From<ExecutionMetadata> for ExecutionMetadataView {
...
+                                // other costs have always been mapped one-to-one
+                                Cost::ActionCost { action_cost_kind: action_cost } => {
+                                    format!("{:?}", action_cost).to_ascii_uppercase()
+                                }
+                                Cost::ExtCost { ext_cost_kind: ext_cost } => {
+                                    format!("{:?}", ext_cost).to_ascii_uppercase()
+                                }
+                                Cost::WasmInstruction => "WASM_INSTRUCTION".to_string(),
+                            },
+                            gas_used: profile_data[cost],
```
Comment (matklad) [inline core/primitives/src/views.rs:1292]: Here, we don't need to go out of our way to preserve behavior, it's ok to add and remove entries to this table.

File: core/primitives-core/src/profile.rs [hunk 4]
```diff
@@ -180,96 +165,38 @@ pub enum Cost {
...
+        // make sure to increase `DataArray::LEN` when adding a new index
         match self {
             Cost::ActionCost { action_cost_kind: ActionCosts::create_account } => 0,
             Cost::ActionCost { action_cost_kind: ActionCosts::delete_account } => 1,
-            Cost::ActionCost { action_cost_kind: ActionCosts::deploy_contract } => 2,
-            Cost::ActionCost { action_cost_kind: ActionCosts::function_call } => 3,
+            Cost::ActionCost { action_cost_kind: ActionCosts::deploy_contract_base } => 2,
+            Cost::ActionCost { action_cost_kind: ActionCosts::deploy_contract_byte } => 2,
+            Cost::ActionCost { action_cost_kind: ActionCosts::function_call_base } => 3,
+            Cost::ActionCost { action_cost_kind: ActionCosts::function_call_byte } => 3,
```
Comment (matklad) [inline core/primitives-core/src/profile.rs:189]: Yeah, I think it's fine to map them to a single entry for the time being. Long term, we should just make the array bigger. The serialization/deserialization code doesn't really care about length, is just zeros unknown cost.

It's better _not_ to change the meaning of indicies, to simplify historical analysis, but even that, strinctly speaking, is not required. If we want, wee can reassign the meaning without protocol change.

File: core/primitives-core/src/config.rs [hunk 5]
```diff
@@ -647,19 +649,27 @@ pub enum ExtCosts {
...
 #[allow(non_camel_case_types)]
 pub enum ActionCosts {
     create_account,
     delete_account,
-    deploy_contract,
-    function_call,
+    deploy_contract_base,
+    deploy_contract_byte,
+    function_call_base,
+    function_call_byte,
```
Comment (matklad) [inline core/primitives-core/src/config.rs:662]: :heart_eyes:

**Review by matklad — APPROVED** [review]

File: core/primitives/src/views.rs (hunk 3 above)
Comment (jakmeier) [reply core/primitives/src/views.rs:1292]: Then what would you suggest? For example, I replace `new_receipt` with `new_action_receipt`, `new_data_receipt_base`, and `new_data_receipt_byte`. The database only has one number. The only way I can see this number make sense it to keep using the old name.

I want to add new entries for the new profile version but that will only happen a few PRs later. And this code here will still remain to handle the old data in the DBs.

Comment (matklad) [reply core/primitives/src/views.rs:1292]: I don't suggest changing anything here, that's just a general note that, when we add two new costs, we can just drop the old one, we don't need to care about stuffing some dummy value there or something

Comment (jakmeier) [reply core/primitives/src/views.rs:1292]: Oh I see, yes that makes sense!

File: core/primitives-core/src/profile.rs (hunk 2 above)
Comment (jakmeier) [reply core/primitives-core/src/profile.rs]: good call! changed it `fold` with saturating addition in the two places I used `sum` before

File: core/primitives-core/src/profile.rs (hunk 1 above)
Comment (jakmeier) [reply core/primitives-core/src/profile.rs:15]: It's not quite `iter().count()`, we have to dedup it first now that multiple costs map to the same entry. But you are right and I added a test to check this.

**Pull request merged**